### DISCUSSION TITLE
I don't understand how root_url is supposed to work

### DISCUSSION THREAD

**Discussion by lahvak** [fonsp/Pluto.jl#1084] [Questions about using Pluto] [answered]
I tried to launch Pluto with
```
Pluto.run(launch_browser=false, require_secret_for_access=false, root_url="blah/")
```
Pluto responds with `Go to blah/ in your browser to start writing ~ have fun!`

If I try to access `http://localhost:1234/blah/`, I get a 404 error, and the server (I do make a connection to the server) says "Not found!".

What am I doing wrong?

**Comment by fonsp** [comment]
What are you trying to do?

**Reply by lahvak** [reply]
I am trying to make Pluto work with [CoCalc](https://cocalc.com/).  Pluto runs on CoCalc, and I can access it from a browser running on CoCalc, but to be able to access it remotely, support for `base_url` is needed, in a similar way as it is implemented in Jupyter.  I was hoping that `root_url` would do that, but in fact I cannot even make `root_url` do anything at all.

**Reply by lahvak** [reply]
Here is the issue I have opened on cocalc: #https://github.com/sagemathinc/cocalc/issues/5270

**Comment by fonsp** [comment, marked as answer]
I think you are looking at this the other way around:

Pluto is designed to work on _any_ base URL, we make sure that all URLs are relative (asset imports, websocket, redirects, links, etc). This means that if you have an nginx proxy set up to serve pluto at, say `runner.cocalc.com/ba9c94/port/pluto/`, then you can just go to that address and everything will work.


The `root_url` setting does not change the server configuration, it only affects the message printed to the terminal, which is probably not relevant to you. When the project started, we needed to specify the `root_url` on custom proxies, just like with Jupyter, but we have found some tricks to avoid it altogether: you do not need to set the `root_url` to run Pluto through a proxy.

**Reply by lahvak** [reply]
I am not sure I understand.  If I run Pluto using `Pluto.run(require_secret_for_access=false, launch_browser=false)`, are you saying that I should be able to access Pluto on any URL in the form `http://localhost:1234/blablah/`?  Because that does not work, it says "Not found!".

I don't think I as a cocalc user have any control over the url coclac will redirect requests to.  I was hoping that I will be able to tell Pluto to use that url with `root_url`, but that does not seem to work.

Is there any way I can set some sort of verbose mode in Pluto, where I would get some information about what url's are trying to connect to it?  Or is there a log file somewhere? Right now I just get "Not found!" in the browser window, and a 404 response in http headers.

**Reply by williamstein** [reply]
@fonsp, I understand what you're saying.  That's very nice that you implemented all of Pluto in such a way that all URL's are relative!

@lahvak the solution on CoCalc is that there is a different mechanism for using a server that does what Pluto does, and it DOES work.  Here's how:

1. Run Pluto via `Pluto.run(launch_browser = false, require_secret_for_access = false, host="0.0.0.0")`.  In particular, you will not even use the `root_url` option at all.
2. In your web browser, visit `https://cocalc.com/[PROJECT_ID]/server/1234/`.
3. See this:
![image](https://user-images.githubusercontent.com/1276278/115065248-c0b65f00-9ea2-11eb-98fa-2ba8f5361a34.png)
4. Click "New Notebook" and do a test and see this:
![image](https://user-images.githubusercontent.com/1276278/115065295-cf9d1180-9ea2-11eb-85d0-d4e47a3a7393.png)

Note that in CoCalc it's necessary right now to install Pluto. To do that:
```sh
export JULIA_DEPOT_PATH=$HOME/julia_depot
$ julia
julia> ]
pkg> add Pluto
# wait a few seconds
```

**Reply by fonsp** [reply]
Just to confirm: what you used, `Pluto.run(launch_browser = false, require_secret_for_access = false, host="0.0.0.0")` is exactly the set of (security) configurations that I would recommend for this setup!

**Comment by williamstein** [comment]
Wow, you guys put the output above the input.  That's pretty intriguing.  :-)  I've implemented a lot of online code environments and never thought of doing that.  I wonder what other things you've done that I would have never thought of in a million years...

![image](https://user-images.githubusercontent.com/1276278/115066061-da0bdb00-9ea3-11eb-9891-f804884e4a2e.png)

**Reply by lahvak** [reply]
Awesome, that's great!  So the practical difference is in using `server` instead of `port` in the URL on CoCalc.  Thanks all of you for the help!

**Reply by fonsp** [reply]
Thanks for the kind message! That particular design choice is not always as well-received, and it's great to hear this response from someone who has worked on similar problems!

I have not used cocalc before, but we are an education-first project, so I would be interested to see Pluto being used on your platform! Let me know if you have any cool ideas on how to better integrate the two, or generally, how to improve the education use case. fons@plutojl.org

**Reply by fonsp** [reply]
One project that might be interesting to you is our recent ability to show a static notebook preview on the web, and connect to a live instance in the background: https://github.com/fonsp/Pluto.jl/pull/811

An example of this being used is https://computationalthinking.mit.edu/

**Reply by williamstein** [reply]
> One project that might be interesting to you is our recent ability to show a static notebook preview on the web, and connect to a live instance in the background: #811

Thanks -- that does interest me.  I wrote something half-way in that direction called the CoCalc Share Server:  

https://share.cocalc.com/share/

It uses a custom lightweight server-side renderer of Jupyter notebooks (and numerous other formats, which could include Pluto at some point), with 1-click to open them in CoCalc (without requiring account creation).  However, there is a lot more to do to make it better (e.g., much faster startup by having a pool of pre-created anonymous accounts, and better support for iframe embedding of static pages).  I hope to get back to this soon, since I haven't really had a chance to work on this seriously in a while.

> I have not used cocalc before, but we are an education-first project, so I would be interested to see Pluto being used on your platform!

Cool.  It's likely I would start by just making it so in CoCalc it's 1-click to launch Pluto in a new tab, running from a cocalc project.  I could also embed a Pluto server for a specific file in an iframe as a tab, like I've also done with Jupyter classic.

What is the situation related to Pluto and realtime synchronization?   In particular, what is the underlying document format and is there a way to access the document embedded in an iframe, and read/write it?   CoCalc has a pretty powerful general realtime sync library to support multiple simultaneous users...

> Thanks for the kind message! That particular design choice is not always as well-received, and it's great to hear this response from someone who has worked on similar problems!

Yes, of course.  Innovation in running code online should be strongly encouraged.   What you've done reminds me a little of what MS Excel does, which is to focus on showing the user the _output_, not the input.  After all, if in a spreadsheet you type `=2+3`, you see 5 in the cell, and only view the expression that leads to that 5 when you specifically put your cursor on the 5 and look at the top of the screen:


![image](https://user-images.githubusercontent.com/1276278/115095366-e14bdc80-9ed5-11eb-99a6-5c84883fb3ff.png)

Have you played around with going even further and not showing the output at all unless the output is selected, just like Excel does?
